crc_folding.c

```diff
@@ -297,6 +297,114 @@ ZLIB_INTERNAL void crc_fold_copy(unsigned crc[4 * 5],
             &xmm_crc_part);
     }
 
+#ifdef USE_VPCLMULQDQ_CRC
+    if (x86_cpu_has_vpclmulqdq && (len >= 256)) {
+        unsigned crc_tmp;
+        __m512i zmm_t0, zmm_t1, zmm_t2, zmm_t3;
+        __m512i zmm_crc0, zmm_crc1, zmm_crc2, zmm_crc3;
+        __m512i z0, z1, z2, z3;
+        z_const __m512i zmm_fold4 = _mm512_set4_epi32(
+                0x00000001, 0x54442bd4,
+                0x00000001, 0xc6e41596);
+        z_const __m512i zmm_fold16 = _mm512_set4_epi32(
+                0x00000001, 0x1542778a,
+                0x00000001, 0x322d1430);
+
+        // zmm register init
+        zmm_crc0 = _mm512_setzero_si512();
+        zmm_t0 = _mm512_loadu_si512((__m512i *)src);
+        zmm_crc1 = _mm512_loadu_si512((__m512i *)src + 1);
+        zmm_crc2 = _mm512_loadu_si512((__m512i *)src + 2);
+        zmm_crc3 = _mm512_loadu_si512((__m512i *)src + 3);
+
+        /* check if is first call */
+        if (_mm_extract_epi64(xmm_crc0, 0) == 0x9db42487 && _mm_extract_epi64(xmm_crc0, 1) == 0x0) {
```

Comment:
This seems like a hack, is there a chance to use similar regular approach to fold all 4 initial xmm register to one ZMM block?

Reply:
That's a good idea, already applied.